pkg/wakaama/include/objects/ipso_sensor_base.h

```diff
@@ -0,0 +1,202 @@
+ *
+ * @{
+ *
+ * @file
+ *
+ * @author      Leandro Lanzieri <leandro.lanzieri@haw-hamburg.de>
+ */
+
+#ifndef OBJECTS_IPSO_SENSOR_BASE_H
+#define OBJECTS_IPSO_SENSOR_BASE_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include "mutex.h"
+#include "liblwm2m.h"
+#include "lwm2m_client.h"
+
+/**
+ * @defgroup lwm2m_objects_ipso_sensor_base_config LwM2M IPSO Sensor base object compile configurations
+ * @ingroup lwm2m_client_config
+ * @{
+ */
+/**
+ * @brief Maximum size for the units resource string
+ */
+#ifndef CONFIG_LWM2M_IPSO_SENSOR_BASE_UNITS_MAX_SIZE
+#define CONFIG_LWM2M_IPSO_SENSOR_BASE_UNITS_MAX_SIZE (4U)
+#endif
+/** @} */
+
+/**
+ * @name IPSO sensor base object resource's IDs (shared among many IPSO sensor objects)
+ * @{
+ */
+/**
+ * @brief Sensor value resource ID.
+ */
+#define LWM2M_IPSO_BASE_SENSOR_VALUE_ID 5700
+/**
+ * @brief Min. measured value resource ID.
+ */
+#define LWM2M_IPSO_BASE_SENSOR_MIN_MEASURED_VALUE_ID 5601
+/**
+ * @brief Max. measured value resource ID.
+ */
+#define LWM2M_IPSO_BASE_SENSOR_MAX_MEASURED_VALUE_ID 5602
+/**
+ * @brief Min. range value resource ID.
+ */
+#define LWM2M_IPSO_BASE_SENSOR_MIN_RANGE_VALUE_ID 5603
+/**
+ * @brief Max. range value resource ID.
+ */
+#define LWM2M_IPSO_BASE_SENSOR_MAX_RANGE_VALUE_ID 5604
+/**
+ * @brief Reset min. and max. measured values resource ID.
+ */
+#define LWM2M_IPSO_BASE_SENSOR_RESET_MIN_MAX_MEASURED_VALUES_ID 5605
+/**
+ * @brief Sensor units resource ID.
+ */
+#define LWM2M_IPSO_BASE_SENSOR_UNITS_ID 5701
+/** @} */
+
+/**
+ * @brief Callback for reading the sensor value.
+ *
+ * @param[in]  read_cb_arg  Data passed for the read callback when the instance was created.
+ * @param[out] value       Pointer to the variable where the value will be stored.
+ *
+ * @return 0 on success
+ * @return <0 otherwise
+ */
+typedef int lwm2m_obj_ipso_base_sensor_read_cb_t(void *read_cb_arg, int16_t *value);
+
+/**
+ * @brief Arguments for the creation of an object based on the IPSO Sensor Base
+ *        object instance.
+ */
+typedef struct lwm2m_obj_ipso_sensor_base_args {
+    int32_t instance_id;                            /**< ID for the new instance. It must be between 0 and (UINT16_MAX - 1),
+                                                         if -1 the next available ID will be used. */
+    char *units;                                    /**< Null-terminated string of the "Sensor Units" resource. May be NULL. */
+    size_t units_len;                               /**< Length of units */
+    int16_t min_range_value;                        /**< Minimum value that can be measured by the sensor. */
+    int16_t max_range_value;                        /**< Maximum value that can be measured by the sensor. */
+    void *read_cb_arg;                              /**< Data to pass to the read callback. May be NULL. */
+    lwm2m_obj_ipso_base_sensor_read_cb_t *read_cb;  /**< Callback to read the sensor value. May be NULL. */
```

Comment:
Why can the call back be null?

Reply:
The idea was to support both push and poll methods. If you want to update your object values whenever there's a new sensor reading, you can use `lwm2m_object_ipso_sensor_base_update_value` (this also avoids triggering a hardware readout by a remote request). If you're fine with executing a readout every time your resource is read, you can use the callback approach (of course, you can also implement this function to use some cached value).